File: reference-impls/python/uacp/extensions/reasoning.py
from __future__ import annotations

VALID_VISIBILITY = {'visible', 'hidden', 'redacted'}
MAX_THINKING_TEXT = 1_000_000


def _err(path: str, code: str, message: str) -> dict:
    return {'path': path, 'code': code, 'message': message}


def validate_reasoning(doc: dict) -> dict:
    errors: list[dict] = []
    messages = doc.get('messages') or []

    for i, m in enumerate(messages):
        if not isinstance(m, dict):
            continue
        content = m.get('content')
        if not isinstance(content, list):
            continue
        for j, block in enumerate(content):
            if not isinstance(block, dict) or block.get('type') != 'thinking':
                continue
            path = f'messages[{i}].content[{j}]'

            text = block.get('text')
            if not isinstance(text, str):
                errors.append(_err(f'{path}.text', 'thinking_missing_text',
                                   'thinking block must have a text field of type string'))
            elif len(text) > MAX_THINKING_TEXT:
                errors.append(_err(f'{path}.text', 'thinking_text_too_long',
                                   f'thinking text must be at most {MAX_THINKING_TEXT} Unicode codepoints'))

            vis = block.get('model_visibility')
            if vis is not None and vis not in VALID_VISIBILITY:
                errors.append(_err(f'{path}.model_visibility', 'model_visibility_invalid',
                                   'model_visibility must be one of: visible, hidden, redacted'))

            tokens = block.get('tokens')
            if tokens is not None:
                if not isinstance(tokens, int) or isinstance(tokens, bool) or tokens < 0:
                    errors.append(_err(f'{path}.tokens', 'tokens_negative',
                                       'tokens must be a non-negative integer'))

    return {'valid': len(errors) == 0, 'errors': errors}

File: reference-impls/go/uacp/extensions/reasoning.go
package extensions

import (
	"fmt"
	"unicode/utf8"
)

const maxThinkingText = 1_000_000

var validVisibility = map[string]bool{
	"visible": true, "hidden": true, "redacted": true,
}

func codepointLen(s string) int {
	return utf8.RuneCountInString(s)
}

// ValidateReasoning enforces the uacp-reasoning extension rules:
// thinking content blocks MUST have a string text field of at most 1_000_000
// codepoints, an optional model_visibility from {visible, hidden, redacted},
// and an optional non-negative integer tokens field.
func ValidateReasoning(doc map[string]any) ExtensionResult {
	errors := []ExtensionError{}
	messagesRaw, _ := doc["messages"].([]any)

	for i, mRaw := range messagesRaw {
		m, ok := mRaw.(map[string]any)
		if !ok {
			continue
		}
		content, ok := m["content"].([]any)
		if !ok {
			continue
		}
		for j, blockRaw := range content {
			block, ok := blockRaw.(map[string]any)
			if !ok {
				continue
			}
			if block["type"] != "thinking" {
				continue
			}
			path := fmt.Sprintf("messages[%d].content[%d]", i, j)

			text, hasText := block["text"].(string)
			if !hasText {
				errors = append(errors, extErr(path+".text", "thinking_missing_text",
					"thinking block must have a text field of type string"))
			} else if codepointLen(text) > maxThinkingText {
				errors = append(errors, extErr(path+".text", "thinking_text_too_long",
					fmt.Sprintf("thinking text must be at most %d Unicode codepoints", maxThinkingText)))
			}

			if visRaw, exists := block["model_visibility"]; exists {
				vis, ok := visRaw.(string)
				if !ok || !validVisibility[vis] {
					errors = append(errors, extErr(path+".model_visibility", "model_visibility_invalid",
						"model_visibility must be one of: visible, hidden, redacted"))
				}
			}

			if tokensRaw, exists := block["tokens"]; exists {
				if n, ok := tokensRaw.(float64); !ok || n < 0 || n != float64(int64(n)) {
					errors = append(errors, extErr(path+".tokens", "tokens_negative",
						"tokens must be a non-negative integer"))
				}
			}
		}
	}

	return ExtensionResult{Valid: len(errors) == 0, Errors: errors}
}
